### DISCUSSION TITLE
EncryptedAssertion not working with ws2007FederationHttpBinding

### DISCUSSION THREAD

**Discussion by eric-git** [CoreWCF/CoreWCF#1478] [Q&A]
I have a "cannot read token" exception, when the CoreWCF-based service is trying to read the token. The SAML 2.0 token is encrypted. 

<blockquote>
fail: Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware[1]
      An unhandled exception has occurred while executing the request.
      CoreWCF.Security.MessageSecurityException: Message security verification failed.
       ---> System.Xml.XmlException: Cannot read the token from the 'EncryptedAssertion' element with the 'urn:oasis:names:tc:SAML:2.0:assertion' namespace for BinarySecretSecurityToken, with a '' ValueType. If this element is expected to be valid, ensure that security is configured to consume tokens with the name, namespace and value type specified.
         at CoreWCF.Security.WSSecurityTokenSerializer.ReadTokenCore(XmlReader reader, SecurityTokenResolver tokenResolver)
         at CoreWCF.Security.WsSecurityTokenSerializerAdapter.ReadTokenCore(XmlReader reader, SecurityTokenResolver tokenResolver)
         at CoreWCF.IdentityModel.Selectors.SecurityTokenSerializer.ReadToken(XmlReader reader, SecurityTokenResolver tokenResolver)
         at CoreWCF.Security.ReceiveSecurityHeader.ReadTokenAsync(XmlReader reader, SecurityTokenResolver tokenResolver, IList`1 allowedTokenAuthenticators)
         at CoreWCF.Security.ReceiveSecurityHeader.ReadTokenAsync(XmlDictionaryReader reader, Int32 position, Byte[] decryptedBuffer, SecurityToken encryptionToken, String idInEncryptedForm, TimeSpan timeout)
         at CoreWCF.Security.ReceiveSecurityHeader.ExecuteFullPassAsync(XmlDictionaryReader reader)
         at CoreWCF.Security.ReceiveSecurityHeader.ProcessAsync(TimeSpan timeout, ChannelBinding channelBinding, ExtendedProtectionPolicy extendedProtectionPolicy)
         at CoreWCF.Security.TransportSecurityProtocol.VerifyIncomingMessageCoreAsync(Message message, TimeSpan timeout)
         at CoreWCF.Security.TransportSecurityProtocol.VerifyIncomingMessageAsync(Message message, TimeSpan timeout)
         --- End of inner exception stack trace ---
         at CoreWCF.Security.TransportSecurityProtocol.VerifyIncomingMessageAsync(Message message, TimeSpan timeout)
         at CoreWCF.Security.SecurityProtocol.VerifyIncomingMessageAsync(Message message, TimeSpan timeout, SecurityProtocolCorrelationState[] correlationStates)
         at CoreWCF.Dispatcher.ServerSecurityChannelDispatcher`1.VerifyIncomingMessageAsync(Message message, TimeSpan timeout, SecurityProtocolCorrelationState[] correlationState)
         at CoreWCF.Dispatcher.SecurityReplyChannelDispatcher.ProcessReceivedRequestAsync(RequestContext requestContext)
         at CoreWCF.Dispatcher.SecurityReplyChannelDispatcher.ProcessReceivedRequestAsync(RequestContext requestContext)
         at CoreWCF.Dispatcher.SecurityReplyChannelDispatcher.DispatchAsync(RequestContext context)
         at CoreWCF.Channels.AspNetCoreReplyChannel.HandleRequest(HttpContext context)
         at CoreWCF.Channels.RequestDelegateHandler.HandleRequest(HttpContext context)
         at CoreWCF.Channels.ServiceModelHttpMiddleware.InvokeAsync(HttpContext context)
         at CoreWCF.Channels.MetadataMiddleware.InvokeAsync(HttpContext context)
         at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddlewareImpl.Invoke(HttpContext context)
</blockquote>

Code:
``` cs
WS2007FederationHttpBinding ws2007FederationHttpBinding = new(WSFederationHttpSecurityMode.TransportWithMessageCredential)
{
    MaxReceivedMessageSize = 2097152
};

serviceBuilder.AddService<Service>(options =>
{
    options.DebugBehavior.IncludeExceptionDetailInFaults = true;
});
serviceBuilder.AddServiceEndpoint<Service, IService>(ws2007FederationHttpBinding, "service.svc");

var message = ws2007FederationHttpBinding.Security.Message;
message.IssuerAddress = new EndpointAddress("https://issuer/service.svc");
message.EstablishSecurityContext = false;
message.IssuedKeyType = SecurityKeyType.SymmetricKey;
message.IssuedTokenType = "http://docs.oasis-open.org/wss/oasis-wss-saml-token-profile-1.1#SAMLV2.0";

var requirements = message.ClaimTypeRequirements;
requirements.Add(new ClaimTypeRequirement("http://uri/one"));
requirements.Add(new ClaimTypeRequirement("http://uri/two"));

serviceBuilder.ConfigureServiceHostBase<IService>(host =>
{
    host.Credentials.ServiceCertificate.SetCertificate(StoreLocation.LocalMachine, StoreName.My, X509FindType.FindByThumbprint, "<thumbprint>");
    host.Credentials.UseIdentityConfiguration = true;

    var identityConfiguration = host.Credentials.IdentityConfiguration;
    identityConfiguration.AudienceRestriction.AudienceMode = AudienceUriMode.Always;
    identityConfiguration.AudienceRestriction.AllowedAudienceUris.Add(new Uri("http://test/service.svc"));
    identityConfiguration.CertificateValidationMode = X509CertificateValidationMode.None;

    ConfigurationBasedIssuerNameRegistry configurationBasedIssuerNameRegistry = new();
    configurationBasedIssuerNameRegistry.AddTrustedIssuer("<thumbprint>", "<sts>");
    identityConfiguration.IssuerNameRegistry = configurationBasedIssuerNameRegistry;
});
```

SOAP request security:
```xml
<o:Security s:mustUnderstand="1" xmlns:o="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd">
	<u:Timestamp u:Id="_0">
		<u:Created>2024-10-18T09:21:26.788Z</u:Created>
		<u:Expires>2024-10-18T09:26:26.788Z</u:Expires>
	</u:Timestamp>
	<saml:EncryptedAssertion xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion">
		<xenc:EncryptedData Id="ED-08f6d921-23a6-42ae-b577-3e93652d09bb" Type="http://www.w3.org/2001/04/xmlenc#Element" xmlns:xenc="http://www.w3.org/2001/04/xmlenc#">
			<xenc:EncryptionMethod Algorithm="http://www.w3.org/2001/04/xmlenc#aes256-cbc"/>
			<ds:KeyInfo xmlns:ds="http://www.w3.org/2000/09/xmldsig#">
				<xenc:EncryptedKey Id="EK-0afbed5a-28dc-4c27-9589-100173790321">
					<xenc:EncryptionMethod Algorithm="http://www.w3.org/2001/04/xmlenc#rsa-oaep-mgf1p">
						<ds:DigestMethod Algorithm="http://www.w3.org/2000/09/xmldsig#sha1"/>
					</xenc:EncryptionMethod>
					<ds:KeyInfo>
						<wsse:SecurityTokenReference xmlns:wsse="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd">
							<ds:X509Data>
								<ds:X509IssuerSerial>
									<ds:X509IssuerName>CN=omitted,OU=omitted,O=omitted,C=AU</ds:X509IssuerName>
									<ds:X509SerialNumber>667127305496342621706403373559713919376684653146</ds:X509SerialNumber>
								</ds:X509IssuerSerial>
							</ds:X509Data>
						</wsse:SecurityTokenReference>
					</ds:KeyInfo>
					<xenc:CipherData>
						<xenc:CipherValue>omitted</xenc:CipherValue>
					</xenc:CipherData>
				</xenc:EncryptedKey>
			</ds:KeyInfo>
			<xenc:CipherData>
				<xenc:CipherValue>omitted</xenc:CipherValue>
			</xenc:CipherData>
		</xenc:EncryptedData>
	</saml:EncryptedAssertion>
	<Signature xmlns="http://www.w3.org/2000/09/xmldsig#">
		<SignedInfo>
			<CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
			<SignatureMethod Algorithm="http://www.w3.org/2000/09/xmldsig#hmac-sha1"/>
			<Reference URI="#_0">
				<Transforms>
					<Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
				</Transforms>
				<DigestMethod Algorithm="http://www.w3.org/2000/09/xmldsig#sha1"/>
				<DigestValue>wQ6Ax5OXddUPHeL8pY6k4moyUIk=</DigestValue>
			</Reference>
		</SignedInfo>
		<SignatureValue>fl26eKlWF1Jryylt5PO+db5+GyA=</SignatureValue>
		<KeyInfo>
			<wsse:SecurityTokenReference a:TokenType="http://docs.oasis-open.org/wss/oasis-wss-saml-token-profile-1.1#SAMLV2.0" xmlns:wsse="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd" xmlns:a="http://docs.oasis-open.org/wss/oasis-wss-wssecurity-secext-1.1.xsd">
				<wsse:KeyIdentifier ValueType="http://docs.oasis-open.org/wss/oasis-wss-saml-token-profile-1.1#SAMLID">_26d26e64-78a2-4ab4-8af0-190f9a8a4900</wsse:KeyIdentifier>
			</wsse:SecurityTokenReference>
		</KeyInfo>
	</Signature>
</o:Security>
```

Is there anything wrong with my code? Or is the message security supported?

@mconnew Can you please give me some advice? I will post more info if needed.

**Comment by mconnew** [comment]
You actually provided all the information I needed. Unfortunately, I don't have good news for you. To handle Saml2, WCF and CoreWCF wrap the `SecurityTokenHandler` api as that's the api model that was developed for newer token types. You see this with the `WsSecurityTokenSerializerAdapter` class on the call stack, it iterates through all the `SecurityTokenHandler` instances to try and handle the token.  
CoreWCF doesn't have its own implementation of a `Saml2SecurityTokenHandler`. Instead, it wraps the `Microsoft.IdentityModel.Tokens.Saml2.Saml2SecurityTokenHandler` implementation from the [Microsoft.IdentityModel.Tokens.Saml nuget package](https://www.nuget.org/packages/Microsoft.IdentityModel.Tokens.Saml). To check if a handler matches, there's a `SecurityTokenHandler.CanReadToken` method that's called, and if it returns true, then that handler is used to read the token. You can see the [.NET Framework implementation here](https://github.com/microsoft/referencesource/blob/master/System.IdentityModel/System/IdentityModel/Tokens/Saml2SecurityTokenHandler.cs#L462-L471). You can see it checks if the current element is either `Assertion` or `EncryptedAssertion`. If you look at the MS.IM.Tokens.Saml [implementation of `CanReadToken`](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/8.1.2/src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.cs#L117-L123), you can see that it only checks for `Assertion` and doesn't look for `EncryptedAssertion`. This means the MS.IM implementation doesn't support encrypted assertions.   
I am incredibly hesitant to port the implementation from .NET Framework for multiple security related reasons. When there's a whole team dedicated to implementations in the identity space, it doesn't make sense to have our own implementation.  
My best suggestion is to [open an issue on the IdentityModel repo](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet) asking for that support. Something like that isn't quick to develop though. If you have a narrow set of possibilities for the encryption mechanism, you could potentially manually decrypt the EncryptedAssertion and replace it with the decrypted Assertion. In this particular case, you would need to look up the certificate based on the X509IssuerSerial, then use that to decrypt the inner CipherValue. This is done with rsa-oaep-mgf1p along with sha1 as the digest method to validate it hasn't been modified. Then that decrypted value is then used to decrypt the outer CipherData using aes256-cbc, this time with no digest method applied. This decrypted data should be the regular Assertion that then can be handled normally. There are other limitations to the MS.IM implementation. If SHA1 is used in the decrypted assertion, then it can't handle validating its signature. There is an extensibility point to do so, and it's on a static property that you modify it. If you go that route and this turns out to be the case, I can provide you the implementation and instructions how to use it for SHA1. You have a final problem though, and that's with the final KeyInfo in the signature block. It uses a SecurityTokenReference to point to the saml Assertion (once decrypted). We don't yet support validating signatures based on an external token reference.
